proposals/NNNN-task-executor-preference.md

```diff
@@ -0,0 +1,830 @@
+  await nonisolatedAsyncFunc()
+}
+
+Task.detached(on: executor) {
+  // starts and runs on the 'executor'
+  await nonisolatedAsyncFunc()
+}
+
+await withDiscardingTaskGroup { group in 
+  group.addTask(on: executor) { 
+    // starts and runs on the 'executor'
+    await nonisolatedAsyncFunc()
+  }
+}
+
+func nonisolatedAsyncFunc() async -> Int {
+  // if the Task has a specific executor preference,
+  // runs on that 'executor' rather than on the default global concurrent executor
+  return 42
+} 
+```
+
+or, for a specific scope using the `withTaskExecutor` method:
+
+```swift
+await withTaskExecutor(executor) { 
+  // if not already running on specified 'executor'
+  // the withTaskExecutor would hop to it, and run this closure on it.
+
+  // task groups
+  await withDiscardingTaskGroup { group in 
+    group.addTask {
+      // starts and runs on the 'executor'
+      await nonisolatedAsyncFunc()
+    }
+  }
+  
+  // async let 
+  async let number = nonisolatedAsyncFunc() // starts and runs on 'executor'
+  await number
+}
+```
+
+Notably, the task executor preference is in effect for the entire structured task hierarchy while running in a task or scope where a task executor preference is set. For example, the following snippet illustrates child tasks created inside of a `withTaskExecutor`.
+
+If a task with such executor preference encounters code which is `isolated` to some specific actor, it would adhere to that requirement and hop to it as expected:
+
+```swift
+let capy: Capybara = Capybara()
+actor Capybara { func eat() {} } 
+
+Task(on: executor) {
+  // starts on 'executor', however...
+  try await capy.eat() // still executes actor isolated code on the actor's executor, as expected
+}
+```
+
+## Detailed design
+
+### Setting task executor preference
+
+A new concept of task executor preference is added to Swift Concurrency tasks. This preference is stored in a task and propagated throughout child tasks (such as ones created by TaskGroups and async let).
+
+The preference can be set using various APIs that will be discussed in detail in their respective sections. The first of those APIs is `withTaskExecutor` which can be called inside an asynchronous context to both ensure we’re executing on the expected executor, as well as set the task executor preference for the duration of the operation closure:
+
+```swift
+await withTaskExecutor(someExecutor) { 
+ // guaranteed to be executing on someExecutor
+}
+```
+
+Once set, the effect of an executor preference is such that a nonisolated func instead of immediately hopping to the global pool, it may hop to the preferred executor, e.g.:
+
+```swift
+nonisolated func tryMe() async { 
+ // ...
+}
+
+let preferredExecutor: SomeConcreteTaskExecutor = ...
+Task(on: preferredExecutor) {
+  // executes on 'preferredExecutor'
+  await tryMe() // tryMe body would execute on 'preferredExecutor'
+}
+
+ await tryMe() // tryMe body would execute on 'default global concurrent executor'
+```
```

Comment:
`await try` seems to get mis-highlighted by GitHub.  It's silly that you have to do this, but could you just rename this function to something that doesn't start with `try`?

Reply:
I'll clean this up post-merge